### ISSUE TITLE
Replace deep/shallow C stack with an explicit user-provided stack check macro

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#231]
Currently Duktape tries to prevent too deep C recursion by keeping track of recursion depth in code paths where arbitrary recursion might occur. A depth limit is imposed for each case, with a RangeError thrown if a recursion limit is hit. There are two sets of recursion limits: deep limits for unix-like platforms and shallow limits tuned roughly for a target with ~64kB stack.

This model does not really work well because Duktape doesn't really know how the stack behaves on the target. Further, even if a target had a stack of certain size, it might be appropriate to impose a stricter limit for resource containment reasons. Making such decisions can only be done by user code.

The current model could be replaced by:

- Removing the current stack depth tracking and recursion limits entirely. Deprecate `DUK_OPT_DEEP_C_STACK`.
- Adding a DUK_USE_STACK_CHECK() macro which Duktape would call at least in important recursion paths. The macro would indicate if there is enough stack space to continue (until the next stack check call); in concrete terms at least a few kilobytes should be available if the macro indicates there's enough space to proceed.

While writing such a macro is a bit awkward, it'd be more accurate than the very rough limits imposed by recursion depth counts. The Duktape internals would be much easier to maintain: as long as there are enough stack check calls in the arbitrary recursion points there'd be no recursion counters etc to carefully keep in sync.

See discussion in #226.

**Comment by svaarala** [comment]
Prototype in:

- https://github.com/svaarala/duktape/compare/user-stack-check-macro

**Comment by fatcerberus** [comment]
Yeah, I noticed the initial implementation doesn't pass any arguments to the macro, which limits its utility.  I imagine it needs to get some sort of offset, so the user can figure out how much stack space is left relative to what's already used.  I think you mentioned a stack pointer/offset earlier, but no idea if that information is even available while still retaining portability (i.e. no inline assembly or any shenanigans like that).

**Comment by svaarala** [comment]
If there are multiple threads the macro (or a function it ultimately calls) needs to be able to somehow compute "stack used so far" and compare it against a safe limit. A running thread can locate its own stack configuration in some systems while in others it may be more awkward.

AFAIK there is no guarantee that a stack is continuous in memory so the check is not necessarily a pointer subtraction; see e.g. https://gcc.gnu.org/wiki/SplitStacks

**Comment by svaarala** [comment]
While I think a stack check would be preferable to Duktape awkwardly tracking recursion depth in multiple places, there may be some previous recursion limits that should be kept. For example, there is a value stack size limit, so perhaps the call stack size limit should also be kept.

**Comment by fatcerberus** [comment]
Ugh, that split stack thing looks like way more trouble and complexity than it's worth.  I don't envy anyone writing a debugger for that. :stuck_out_tongue_closed_eyes: 

Overall though I agree with the rationale of this change.  Plus getting rid of most the recursion counters may provide a speed increase too!

**Comment by svaarala** [comment]
I don't think the speed increase will be really measurable, unfortunately :)

But the counters are awkward to maintain and need to be implemented whenever unbounded recursion is possible. It's easy to miss such places, and the limits are individual to each mechanism so the limits for a certain stack size need to be considered together and can easily be overly conservative, etc. I don't think the previous mechanism worked very well for anyone so it's probably good to consider some alternatives. Not sure if this is it though... :)

**Comment by svaarala** [comment]
Considering the environments I've worked with, I'd be mostly happy with a call stack limit to catch unintended user code recursion (this wouldn't catch tail calls of course), and no explicit stack checks.

There's one interesting environment with a stack size smaller than 64kB and stack overruns resulting in a hard reboot (very painful to diagnose). For that environment a stack check would be very useful.

**Comment by fatcerberus** [comment]
"Stack overruns resulting in a hard reboot"

Ouch.  Reminds me of my MS-DOS programming days.  Bugs that cause a harmless segfault nowadays usually locked up the machine in real mode.  I quickly became very familiar with Ctrl+Alt+Del. :)
